File: medium/3702.py
"""
3702. Longest Subsequence With Non-Zero Bitwise XOR

https://leetcode.com/problems/longest-subsequence-with-non-zero-bitwise-xor/

Weekly Contest 470
"""

from functools import reduce
from typing import List


class Solution:
    def longestSubsequence(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return 1 if nums[0] != 0 else 0

        all_xor: int = reduce(lambda a, b: a ^ b, nums)

        if all_xor != 0:
            return len(nums)

        if all(num == 0 for num in nums):
            return 0

        return len(nums) - 1

File: medium/3702.rs
// 3702. Longest Subsequence With Non-Zero Bitwise XOR
// https://leetcode.com/problems/longest-subsequence-with-non-zero-bitwise-xor/
// Weekly Contest 514

impl Solution {
    pub fn longest_subsequence(nums: Vec<i32>) -> i32 {
        let all_xor: i32 = nums.iter().fold(0, |acc, num| acc ^ num);

        if all_xor != 0 {
            return nums.len() as i32;
        }

        if nums.iter().all(|num| num == &0) {
            return 0;
        }

        nums.len() as i32 - 1
    }
}
